### PR TITLE
Update wayland-client version to 0.30

### LINKED ISSUE #800: Update to wayland-client 0.30
CC @GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #800.

I might need your help @bilelmoussaoui, I currently have this error (multiple times):

```
error[E0599]: no method named `display` found for reference `&WaylandDevice` in the current scope
  --> gdkwayland/src/wayland_device.rs:35:28
   |
35 |         let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
   |                            ^^^^^^^ method not found in `&WaylandDevice`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `display`, perhaps you need to implement it:
           candidate #1: `AppLaunchContextExt`
```

But the objects inherit from others which do have this method, so I'm a bit lost. Any idea?

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
I believe the issue is you lack a `use gdk::prelude::*;`

**Comment by GuillaumeGomez** [comment]
@bilelmoussaoui I think I maybe forgot to add something in `gdk` since when I add the missing import we more or less have the same issue:

```
error[E0599]: the method `display` exists for reference `&WaylandDevice`, but its trait bounds were not satisfied
  --> gdkwayland/src/wayland_device.rs:36:28
   |
11 | / glib::wrapper! {
12 | |     #[doc(alias = "GdkWaylandDevice")]
13 | |     pub struct WaylandDevice(Object<ffi::GdkWaylandDevice>) @extends gdk::Device;
14 | |
...  |
17 | |     }
18 | | }
   | | -
   | | |
   | |_doesn't satisfy `WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
   |   doesn't satisfy `WaylandDevice: gdk::prelude::AppLaunchContextExt`
...
36 |           let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
   |                              ^^^^^^^ method cannot be called on `&WaylandDevice` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
           which is required by `WaylandDevice: gdk::prelude::AppLaunchContextExt`
           `&WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
           which is required by `&WaylandDevice: gdk::prelude::AppLaunchContextExt`
note: the trait `IsA` must be implemented
  --> /home/imperio/.cargo/git/checkouts/gtk-rs-core-7be42ca38bd6361c/e69aa51/glib/src/object.rs:65:1
   |
65 | / pub unsafe trait IsA<T: ObjectType>:
66 | |     ObjectType + Into<T> + AsRef<T> + std::borrow::Borrow<T>
   | |____________________________________________________________^

error[E0599]: the method `display` exists for reference `&WaylandDevice`, but its trait bounds were not satisfied
  --> gdkwayland/src/wayland_device.rs:53:28
   |
11 | / glib::wrapper! {
12 | |     #[doc(alias = "GdkWaylandDevice")]
13 | |     pub struct WaylandDevice(Object<ffi::GdkWaylandDevice>) @extends gdk::Device;
14 | |
...  |
17 | |     }
18 | | }
   | | -
   | | |
   | |_doesn't satisfy `WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
   |   doesn't satisfy `WaylandDevice: gdk::prelude::AppLaunchContextExt`
...
53 |           let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
   |                              ^^^^^^^ method cannot be called on `&WaylandDevice` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
           which is required by `WaylandDevice: gdk::prelude::AppLaunchContextExt`
           `&WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
           which is required by `&WaylandDevice: gdk::prelude::AppLaunchContextExt`
note: the trait `IsA` must be implemented
  --> /home/imperio/.cargo/git/checkouts/gtk-rs-core-7be42ca38bd6361c/e69aa51/glib/src/object.rs:65:1
   |
65 | / pub unsafe trait IsA<T: ObjectType>:
66 | |     ObjectType + Into<T> + AsRef<T> + std::borrow::Borrow<T>
   | |____________________________________________________________^

error[E0599]: the method `display` exists for reference `&WaylandDevice`, but its trait bounds were not satisfied
  --> gdkwayland/src/wayland_device.rs:71:28
   |
11 | / glib::wrapper! {
12 | |     #[doc(alias = "GdkWaylandDevice")]
13 | |     pub struct WaylandDevice(Object<ffi::GdkWaylandDevice>) @extends gdk::Device;
14 | |
...  |
17 | |     }
18 | | }
   | | -
   | | |
   | |_doesn't satisfy `WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
   |   doesn't satisfy `WaylandDevice: gdk::prelude::AppLaunchContextExt`
...
71 |           let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
   |                              ^^^^^^^ method cannot be called on `&WaylandDevice` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
           which is required by `WaylandDevice: gdk::prelude::AppLaunchContextExt`
           `&WaylandDevice: IsA<gdk::gio::AppLaunchContext>`
           which is required by `&WaylandDevice: gdk::prelude::AppLaunchContextExt`
note: the trait `IsA` must be implemented
  --> /home/imperio/.cargo/git/checkouts/gtk-rs-core-7be42ca38bd6361c/e69aa51/glib/src/object.rs:65:1
   |
65 | / pub unsafe trait IsA<T: ObjectType>:
66 | |     ObjectType + Into<T> + AsRef<T> + std::borrow::Borrow<T>
   | |____________________________________________________________^

error[E0599]: the method `display` exists for reference `&WaylandMonitor`, but its trait bounds were not satisfied
  --> gdkwayland/src/wayland_monitor.rs:23:28
   |
10 | / glib::wrapper! {
11 | |     #[doc(alias = "GdkWaylandMonitor")]
12 | |     pub struct WaylandMonitor(Object<ffi::GdkWaylandMonitor>) @extends gdk::Monitor;
13 | |
...  |
16 | |     }
17 | | }
   | | -
   | | |
   | |_doesn't satisfy `WaylandMonitor: IsA<gdk::gio::AppLaunchContext>`
   |   doesn't satisfy `WaylandMonitor: gdk::prelude::AppLaunchContextExt`
...
23 |           let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
   |                              ^^^^^^^ method cannot be called on `&WaylandMonitor` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `WaylandMonitor: IsA<gdk::gio::AppLaunchContext>`
           which is required by `WaylandMonitor: gdk::prelude::AppLaunchContextExt`
           `&WaylandMonitor: IsA<gdk::gio::AppLaunchContext>`
           which is required by `&WaylandMonitor: gdk::prelude::AppLaunchContextExt`
note: the trait `IsA` must be implemented
  --> /home/imperio/.cargo/git/checkouts/gtk-rs-core-7be42ca38bd6361c/e69aa51/glib/src/object.rs:65:1
   |
65 | / pub unsafe trait IsA<T: ObjectType>:
66 | |     ObjectType + Into<T> + AsRef<T> + std::borrow::Borrow<T>
   | |____________________________________________________________^

error[E0599]: the method `display` exists for reference `&WaylandSeat`, but its trait bounds were not satisfied
  --> gdkwayland/src/wayland_seat.rs:23:28
   |
10 | / glib::wrapper! {
11 | |     #[doc(alias = "GdkWaylandSeat")]
12 | |     pub struct WaylandSeat(Object<ffi::GdkWaylandSeat>) @extends gdk::Seat;
13 | |
...  |
16 | |     }
17 | | }
   | | -
   | | |
   | |_doesn't satisfy `WaylandSeat: IsA<gdk::gio::AppLaunchContext>`
   |   doesn't satisfy `WaylandSeat: gdk::prelude::AppLaunchContextExt`
...
23 |           let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
   |                              ^^^^^^^ method cannot be called on `&WaylandSeat` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `WaylandSeat: IsA<gdk::gio::AppLaunchContext>`
           which is required by `WaylandSeat: gdk::prelude::AppLaunchContextExt`
           `&WaylandSeat: IsA<gdk::gio::AppLaunchContext>`
           which is required by `&WaylandSeat: gdk::prelude::AppLaunchContextExt`
note: the trait `IsA` must be implemented
  --> /home/imperio/.cargo/git/checkouts/gtk-rs-core-7be42ca38bd6361c/e69aa51/glib/src/object.rs:65:1
   |
65 | / pub unsafe trait IsA<T: ObjectType>:
66 | |     ObjectType + Into<T> + AsRef<T> + std::borrow::Borrow<T>
   | |____________________________________________________________^

warning: unused import: `gdk::prelude`
 --> gdkwayland/src/wayland_device.rs:4:5
  |
4 | use gdk::prelude::*;
  |     ^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `gdk::prelude`
 --> gdkwayland/src/wayland_monitor.rs:3:5
  |
3 | use gdk::prelude::*;
  |     ^^^^^^^^^^^^

warning: unused import: `gdk::prelude`
 --> gdkwayland/src/wayland_seat.rs:3:5
  |
3 | use gdk::prelude::*;
  |     ^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0599`.
warning: `gdkwayland` (lib) generated 3 warnings
error: could not compile `gdkwayland` due to 5 previous errors; 3 warnings emitted
```

**Comment by GuillaumeGomez** [comment]
Found the problem: some types were not marked as "non-final types" so I fixed it.

File: gdk/Gir.toml [hunk 1]
```diff
@@ -421,6 +425,11 @@ status = "generate"
     # the actual method getter after auto-renaming.
     rename = "screen_width_mm"
 
+[[object]]
+name = "Gdk.Seat"
+status = "generate"
+final_type = false
```
Comment (sdroege) [inline gdk/Gir.toml:431]: Yeah the heuristics are going to work better in the new code generator so you don't need to manually set this anymore in most cases

File: gdk/Gir.toml [hunk 2]
```diff
@@ -240,6 +238,7 @@ status = "generate"
 [[object]]
 name = "Gdk.Device"
 status = "generate"
+final_type = false
```
Comment (sdroege) [inline gdk/Gir.toml:241]: You need to move things in `src/device.rs` into a new `DeviceExtManual` trait now

File: gdkwayland/src/wayland_device.rs [hunk 3]
```diff
@@ -31,28 +32,53 @@ impl WaylandDevice {
 
     #[doc(alias = "gdk_wayland_device_get_wl_seat")]
     #[doc(alias = "get_wl_seat")]
-    pub fn wl_seat(&self) -> WlSeat {
+    pub fn wl_seat(&self) -> Option<WlSeat> {
+        let display = self.display().downcast::<crate::WaylandDisplay>().unwrap();
```
Comment (sdroege) [inline gdkwayland/src/wayland_device.rs:36]: You could `unsafe_cast()` here to reduce the amount of generated code

Comment (sdroege) [reply gdkwayland/src/wayland_device.rs:36]: (also elsewhere)

Comment (sdroege) [reply gdkwayland/src/wayland_device.rs:36]: Also you only need the display if the C function below returns non-`NULL` so you can do it afterwards

**Review by sdroege — COMMENTED** [review]
Looks good to me otherwise :)

File: gdk/Gir.toml (hunk 1 above)
Comment (GuillaumeGomez) [reply gdk/Gir.toml:431]: We'll see once it's out. ;)

**Comment by GuillaumeGomez** [comment]
Updated!

File: gdk/Gir.toml (hunk 1 above)
Comment (sdroege) [reply gdk/Gir.toml:431]: Yeah it will take a while, but that one is easy at least

File: gdk/Gir.toml (hunk 2 above)
Comment (GuillaumeGomez) [reply gdk/Gir.toml:241]: I forgot to do that.

**Review by sdroege — APPROVED** [review]

**Pull request merged**